mdsuite/graph_modules/molecular_graph.py

```diff
@@ -247,14 +256,224 @@ def check_a_in_b(a, b):
+                if i == 0:
+                    lengths.append(length)
+                else:
+                    lengths.append(length + lengths[i - 1])
+
+            for i, particle_species in enumerate(self.species):
+                if i == 0:
+                    indices_dict[particle_species] = np.sort(
+                        np.array(list(filter(lambda x: x < lengths[i], graph_dict[item])))
+                    ).tolist()
+                else:
+                    greater_array = list(
+                        filter(lambda x: x >= lengths[i - 1], graph_dict[item])
+                    )
+                    constrained_array = list(
+                        filter(lambda x: x < lengths[i], greater_array)
+                    )
+                    indices_dict[particle_species] = np.sort(
+                        np.array(constrained_array) - (lengths[i - 1])
+                    ).tolist()
+
+            particle_groups[item] = indices_dict
+
+        return particle_groups
+
+
+def build_smiles_graph(smiles_string: str) -> tuple:
+    """
+    Build molecular graphs from SMILES strings.
+
+    Parameters
+    ----------
+    smiles_string : str
+            SMILES string to use in the graph construction.
+
+    Returns
+    -------
+    smiles_graph :
+            Graph object returned by PySmiles
+    species : dict
+            A dict object containing species information about the molecule.
+    """
+    mol = read_smiles(smiles_string, explicit_hydrogen=True)
+    data = mol.nodes
+    species = {}
+    for i in range(len(data)):
+        item = data[i].get("element")
+        if item in species:
+            species[item] += 1
+        else:
+            species[item] = 0
+            species[item] += 1
+
+    return mol, species
+
+
+def _apply_system_cutoff(input_tensor: tf.Tensor, cutoff: float) -> tf.Tensor:
+    """
+    Enforce a cutoff on a tensor.
+
+    In this context the cutoff is used to identify bonded atoms. We argue
+    that the closest atoms will be bonded as defined by a cutoff.
+    Constructing the mask of closest atoms will allow for the bonded ones to be
+    identified.
+
+    Parameters
+    ----------
+    input_tensor : tf.Tensor
+            Tensor of any size or shape to be masked. In our case it is a distance
+            tensor of the atoms in a configuration.
+    cutoff : float
+            Cutoff to use in the mask. If a distance is greater than this cutoff it
+            is marked as 0, if not, it is 1.
+
+    Returns
+    -------
+    masked_tensor : tf.Tensor
+            A tensor of ones and zeros where 1s corresponded to 'bonded' particles
+            and 0s indicated no bonding. Note, the diagonals of this tensor are
+            set to 0 as a particle cannot bond itself.
+    """
+
+    cutoff_mask = tf.cast(
+        tf.less(input_tensor, cutoff), dtype=tf.int16
+    )  # Construct the mask
+
+    return tf.linalg.set_diag(cutoff_mask, np.zeros(len(input_tensor)))
```

Comment:
Possibly `np.zeros_like(input_tensor)` ?

Reply:
Good idea, I can update it